File: mte-sdr-blazor/MteSDRTest/MteSDRTest.Server/Helpers/MteHelper.cs
// ***********************************************************************
// Assembly         : MteSDRTest.Server
// Author           : Eclypses Developer
// Created          : 07-20-2022
//
// Last Modified By : Eclypses Developer
// Last Modified On : 07-20-2022
// ***********************************************************************
// <copyright file="MteHelper.cs" company="MteSDRTest.Server">
//     Copyright (c) Eclypses Inc.. All rights reserved.
// </copyright>
// <summary></summary>
// ***********************************************************************

using System;
using System.Text;
using System.Threading.Tasks;
using Eclypses.MTE;
using Microsoft.Extensions.Logging;

namespace MteSDRTest.Server.Helpers {
    /// <summary>
    /// Class MteHelper.
    /// Implements the <see cref="MteSDRTest.Server.Helpers.IMteHelper" />
    /// </summary>
    /// <seealso cref="MteSDRTest.Server.Helpers.IMteHelper" />
    public class MteHelper : IMteHelper {
        /// <summary>
        /// The logger.
        /// </summary>
        private readonly ILogger<MteHelper> _logger;

        /// <summary>
        /// The cache helper - this wraps the IDistributedCache that holds MTE state.
        /// </summary>
        private readonly ICacheHelper _cache;

        /// <summary>
        /// Gets or sets the mte base.
        /// </summary>
        /// <value>The mte base.</value>
        private static MteBase _mteBase { get; set; }

        /// <summary>
        /// The encoder lock - MTE is single threaded, so throw a lock around its use.
        /// </summary>
        private static object _encoderLock = new object();

        /// <summary>
        /// The decoder lock - MTE is single threaded, so throw a lock around its use.
        /// </summary>
        private static object _decoderLock = new object();

        /// <summary>
        /// Initializes a new instance of the <see cref="MteHelper" /> class.
        /// </summary>
        /// <param name="logger">The logger.</param>
        /// <param name="cache">The cache.</param>
        public MteHelper(ILogger<MteHelper> logger, ICacheHelper cache) {
            _logger = logger;
            _cache = cache;
        }

        #region Instantiate

        /// <summary>
        /// Instantiates a single use MTE for the specified licensed company.
        /// This MUST be called before any other MTE methods can be used, usually as soon as the website loads.
        /// </summary>
        /// <param name="licensedCompany">The licensed company.</param>
        /// <param name="licenseKey">The license key.</param>
        /// <returns><c>true</c> if successful, <c>false</c> otherwise.</returns>
        /// <exception cref="System.ApplicationException">Error instantiating MTE: {_mteBase.GetStatusName(licenseStatus)} - {_mteBase.GetStatusDescription(licenseStatus)}</exception>
        public bool Instantiate(string licensedCompany, string licenseKey) {
            try {
                //
                // Assign _mteBase variable so we can access mte static values.
                //
                _mteBase = new MteBase();
                if (!_mteBase.InitLicense(licensedCompany, licenseKey)) {
                    var licenseStatus = MteStatus.mte_status_license_error;
                    throw new ApplicationException($"Error instantiating MTE: {_mteBase.GetStatusName(licenseStatus)} - {_mteBase.GetStatusDescription(licenseStatus)}");
                }

                return true;
            } catch (Exception ex) {
                _logger.LogError(ex, "Exception - MteBase is not valid.");
                throw;
            }
        }
        #endregion

        #region CreateMteDecoder

        /// <summary>
        /// Creates the MKE decoder. Since we are pairing with a Java Script client,
        /// We must use a string value for entropy (not a byte array) and it should
        /// be used as a series of byte values of the string, not a decoded Base-64 string.
        /// That is because when WASM creates it's entropy, it needs a string and just
        /// converts it to bytes, so we must do the same on the server.
        /// </summary>
        /// <param name="clientId">The client identifier - used to get the proper MTE state.</param>
        /// <param name="personalization">The personalization token to use for this decoder.</param>
        /// <param name="nonce">The nonce value to use for this decoder.</param>
        /// <param name="entropy">The entropy value to use for this decoder.</param>
        /// <returns>MteDec.</returns>
        /// <exception cref="System.ApplicationException">Error instantiating MKE Decoder: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}</exception>
        public async Task CreateMteDecoder(string clientId, string personalization, ulong nonce, string entropy) {
            try {
                //
                // Uncomment to assist with debugging
                // _logger.LogDebug($"Decoder Personalization: {personalization}");
                // _logger.LogDebug($"Decoder Nonce: {nonce}");
                // _logger.LogDebug($"Decoder Entropy: {entropy}");
                //
                string state;
                lock (_decoderLock) {
                    var mteDecoder = new MteDec(sWindow: -63);
                    mteDecoder.SetEntropy(Encoding.UTF8.GetBytes(entropy));
                    mteDecoder.SetNonce(nonce);

                    var status = mteDecoder.Instantiate(personalization);
                    if (status != MteStatus.mte_status_success) {
                        throw new ApplicationException($"Error instantiating MKE Decoder: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}");
                    }

                    state = mteDecoder.SaveStateB64();

                    //
                    // Uncomment to assist with debugging
                    // _logger.LogDebug($"API - Decoder state after instantiation: {state}");
                    //
                    mteDecoder.Uninstantiate();
                    mteDecoder = null;
                }

                await _cache.StoreDecoderCache(state, clientId);
            } catch (Exception ex) {
                _logger.LogError(ex, "Exception - MkeDecoder is not valid.");
                throw;
            }
        }
        #endregion

        #region CreateMteEncoder

        /// <summary>
        /// Creates the MKE encoder. Since we are pairing with a Java Script client,
        /// We must use a string value for entropy (not a byte array) and it should
        /// be used as a series of byte values of the string, not a decoded Base-64 string.
        /// That is because when WASM creates it's entropy, it needs a string and just
        /// converts it to bytes, so we must do the same on the server.
        /// </summary>
        /// <param name="clientId">The client identifier - used to get the proper MTE state.</param>
        /// <param name="personalization">The personalization token to use for this encoder.</param>
        /// <param name="nonce">The nonce value to use for this encoder.</param>
        /// <param name="entropy">The entropy value to use for this encoder.</param>
        /// <returns>MteEnc.</returns>
        /// <exception cref="System.ApplicationException">Error instantiating MKE Encoder: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}</exception>
        public async Task CreateMteEncoder(string clientId, string personalization, ulong nonce, string entropy) {
            try {
                //
                // Uncomment to assist with debugging
                // _logger.LogDebug($"Encoder Personalization: {personalization}");
                // _logger.LogDebug($"Encoder Nonce: {nonce}");
                // _logger.LogDebug($"Encoder Entropy: {entropy}");
                //
                string state;
                lock (_encoderLock) {
                    var mteEncoder = new MteEnc();
                    mteEncoder.SetEntropy(Encoding.UTF8.GetBytes(entropy));
                    mteEncoder.SetNonce(nonce);
                    var status = mteEncoder.Instantiate(personalization);
                    if (status != MteStatus.mte_status_success) {
                        throw new ApplicationException($"Error instantiating MKE Encoder: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}");
                    }

                    state = mteEncoder.SaveStateB64();

                    //
                    // Uncomment to assist with debugging
                    // _logger.LogDebug($"API - Encoder state after instantiation: {state}");
                    //
                    mteEncoder.Uninstantiate();
                    mteEncoder = null;
                }

                await _cache.StoreEncoderCache(state, clientId);
            } catch (Exception ex) {
                _logger.LogError(ex, "Exception - MkeEncoder is not valid.");
                throw;
            }
        }
        #endregion

        #region MkeEncryptString

        /// <summary>
        /// Uses the MkeEncoder to encrypt a clear payload and return it as the encoded string.
        /// </summary>
        /// <param name="clientId">The client identifier.</param>
        /// <param name="clear">The clear string to encode.</param>
        /// <returns>Task&lt;System.Byte[]&gt; of the encoded string.</returns>
        /// <exception cref="System.ApplicationException">Error Encoding bytes: {_mteBase!.GetStatusName(status)} - {_mteBase!.GetStatusDescription(status)}</exception>
        public async Task<string> MkeEncryptString(string clientId, string clear) {
            try {
                var state = await _cache.TakeEncoderCache(clientId);
                string encoded;
                lock (_encoderLock) {
                    //
                    // Create a new encoder and re-hydrate the state that we retrieve from cache for this client.
                    //
                    var mkeEncoder = new MteMkeEnc();

                    mkeEncoder.RestoreStateB64(state);

                    MteStatus status = MteStatus.mte_status_success;
                    encoded = mkeEncoder!.EncodeB64(clear, out status);
                    if (status != MteStatus.mte_status_success) {
                        throw new ApplicationException($"Error Encoding bytes: {_mteBase!.GetStatusName(status)} - {_mteBase!.GetStatusDescription(status)}");
                    }

                    state = mkeEncoder!.SaveStateB64();
                    mkeEncoder!.Uninstantiate();
                    mkeEncoder = null;
                }

                await _cache.StoreEncoderCache(state!, clientId);
                return encoded;
            } catch (Exception ex) {
                _logger.LogError(ex, "Exception - could not encode.");
                throw;
            }
        }
        #endregion

        #region MkeDecryptString

        /// <summary>
        /// Uses the MkeDecoder to decrypt an encoded payload and return it as the clear string.
        /// </summary>
        /// <param name="clientId">The client identifier.</param>
        /// <param name="encoded">The encoded string.</param>
        /// <returns>Task&lt;System.Byte[]&gt; of the original clear string.</returns>
        /// <exception cref="System.ApplicationException">Error Decoding byte array: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}</exception>
        public async Task<string> MkeDecryptString(string clientId, string encoded) {
            try {
                var state = await _cache.TakeDecoderCache(clientId);
                string clear;
                lock (_decoderLock) {
                    var mkeDecoder = new MteMkeDec(sWindow: -63);

                    mkeDecoder.RestoreStateB64(state);
                    MteStatus status = MteStatus.mte_status_success;
                    clear = mkeDecoder!.DecodeStrB64(encoded, out status);
                    if (status != MteStatus.mte_status_success) {
                        throw new ApplicationException($"Error Decoding byte array: {_mteBase!.GetStatusName(status)} - {_mteBase.GetStatusDescription(status)}");
                    }

                    state = mkeDecoder!.SaveStateB64();
                    mkeDecoder!.Uninstantiate();
                    mkeDecoder = null;
                }

                await _cache.StoreDecoderCache(state!, clientId);
                return clear;
            } catch (Exception ex) {
                _logger.LogError(ex, "Exception - could not decode.");
                throw;
            }
        }
        #endregion
    }
}

File: mte-sdr-blazor/MteSDRTest/MteSDRTest.Client/wwwroot/js/mte-helper.js
import {
    MteBase,
    MteWasm,
    MteSdrStorage,
    MteStatus,
    MteMkeEnc,
    MteMkeDec,
} from "./Mte.js";
/**
 * Declare some variables that are global to this module.
 */
let mteWasm = null;
let mteBase = null;
let mtePersistedSdr = null;
let mteSessionSdr = null;

let encoderState = null;
let decoderState = null;
const licenseCompany = "Eclypses Inc.";
const licenseKey = "Eclypses123";
/**
 * An asynchronous function that instantiates MteWasm, then sets up MteBase for future use.
 * This MUST be called before any other MTE methods can be used, usually as soon as the website loads.
 */
export async function instantiateMteWasm() {
    // assign mteWasm variable, and instantiate wasm
    mteWasm = new MteWasm();
    await mteWasm.instantiate();

    // assign mteBase variable
    mteBase = new MteBase(mteWasm);

    // If applicable, Initialize MTE license
    if (!mteBase.initLicense(licenseCompany, licenseKey)) {
        const licenseStatus = MteStatus.mte_status_license_error;
        const status = mteBase.getStatusName(licenseStatus);
        const message = mteBase.getStatusDescription(licenseStatus);
        throw new Error(`Error with MTE License.\n${status}: ${message}`);
    }

    return true;
}

/**
 * Initializes a local MteSdr
 * to use for securely storing and retrieving
 * data items from the browser storage.
 * @param {any} category This allows for specifying what kind of data you with to manage.
 * @param {any} entropy This is a byte array of secure values to use for entropy.
 * @param {any} nonce This is a ulong used to instantiate the MTE.
 * @param {any} persist If this is true, the data is stored in LocalStorage, otherwise it is kept in Session Storage.
 */
function initializeSdr(category, entropy, nonce, persist) {
    const mteSdr = MteSdrStorage.fromdefault(mteWasm, category, persist);
    mteSdr.initSdr(entropy, nonce);
    return mteSdr;
}

/**
 * Initializes a LocalStorage SDR for persisted data between sessions.
 * @param {any} category This allows for specifying what kind of data you with to manage.
 * @param {any} entropy This is a byte array of secure values to use for entropy.
 * @param {any} nonce This is a ulong used to instantiate the MTE.
 */
export async function initializePersistentSdr(category, entropy, nonce) {
    mtePersistedSdr = initializeSdr(category, entropy, nonce, true);
}

/**
 * Initializes a SessionStorage SDR for data only kept within this browser session.
 * @param {any} category This allows for specifying what kind of data you with to manage.
 * @param {any} entropy This is a byte array of secure values to use for entropy.
 * @param {any} nonce This is a ulong used to instantiate the MTE.
 */
export async function initializeSessionSdr(category, entropy, nonce) {
    mteSessionSdr = initializeSdr(category, entropy, nonce, false);
}

/**
 * Retrieves, reveals, and returns the requested data item.
 * @param {any} name The name associated with the item you with to retrieve.
 * @param {any} persistent If this is true, the data is retrieved from LocalStorage, otherwise it is retrieved from Session Storage.
 */
export async function read(name, persistent) {
    try {
        if (persistent) {
            return mtePersistedSdr.readString(name);
        } else {
            return mteSessionSdr.readString(name);
        }
    } catch (err) {
        return null;
    }
}

/**
 * Conceals and stores the requested data item.
 * @param {any} name The name associated with the item you with to store.
 * @param {any} persistent If this is true, the data is stored in LocalStorage, otherwise it is kept in Session Storage.
 */
export async function write(name, data, persistent) {
    if (persistent) {
        mtePersistedSdr.writeString(name, data);
    } else {
        mteSessionSdr.writeString(name, data);
    }
}

/**
 * Removes the requested data item from storage.
 * @param {any} name The name associated with the item you with to remove.
 * @param {any} persistent If this is true, the data is removed from LocalStorage, otherwise it is removed from Session Storage.
 */
export async function remove(name, persistent) {
    if (persistent) {
        mtePersistedSdr.remove(name);
    } else {
        mteSessionSdr.remove(name);
    }
}

/**
 * Initialize MTE encoder/decoder and saves their states
 * @param {string} personalization The personalization string
 * @param {string} nonce A string of integers
 * @param {Uint8Array} encoderEntropy A Uint8Array for the encoder entropy
 * @param {Uint8Array} decoderEntropy A Uint8Array for the decoder entropy
 */
export function createInitialMteState(
    encoderEntropy,
    decoderEntropy,
    nonce,
    personalization
) {
    // create encoder
    const mteEncoder = MteMkeEnc.fromdefault(mteWasm);
    mteEncoder.setEntropyStr(encoderEntropy);
    mteEncoder.setNonce(nonce);
    const encoderStatus = mteEncoder.instantiate(personalization);
    if (encoderStatus !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(encoderStatus);
        const message = mteBase.getStatusDescription(encoderStatus);
        throw new Error(`Error instantiating MTE Encoder.\n${status}: ${message}`);
    }
    // save mte encoder state
    encoderState = mteEncoder.saveStateB64();
    // destroy encoder
    mteEncoder.uninstantiate();
    mteEncoder.destruct();

    // create decoder
    const mteDecoder = MteMkeEnc.fromdefault(mteWasm);
    mteDecoder.setEntropyStr(decoderEntropy);
    mteDecoder.setNonce(nonce);
    const decoderStatus = mteDecoder.instantiate(personalization);
    if (decoderStatus !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(decoderStatus);
        const message = mteBase.getStatusDescription(decoderStatus);
        throw new Error(`Error instantiating MTE Decoder.\n${status}: ${message}`);
    }
    // save mte decoder state
    decoderState = mteDecoder.saveStateB64();
    // destroy decoder
    mteDecoder.uninstantiate();
    mteDecoder.destruct();
}

/**
 * Use MKE to encrypt a uint8Array
 * - create mte encoder, restore it from saved mte state
 * - encode the array
 * - save new encoder state, destroy encoder
 * - return encrypted result
 */
export function mkeEncryptUint8Array(arr) {
    const encoder = MteMkeEnc.fromdefault(mteWasm);
    encoder.restoreStateB64(encoderState);
    const encodeResult = encoder.encode(arr);
    if (encodeResult.status !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(encodeResult.status);
        const message = mteBase.getStatusDescription(encodeResult.status);
        throw new Error(`Error when MKE encrypting data.\n${status}: ${message}`);
    }
    encoderState = encoder.saveStateB64();
    encoder.uninstantiate();
    encoder.destruct();
    return encodeResult.arr;
}

/**
 * Uses the MkeEncoder to encode a string and return it as base64
 * @param str A string value to MTE encode.
 * @return An MTE encoded value as a base64 encoded string.
 */
export function mkeEncryptString(str) {
    const encoder = MteMkeEnc.fromdefault(mteWasm);
    encoder.restoreStateB64(encoderState);
    const encodeResult = encoder.encodeStrB64(str);
    // check if encode failed
    if (encodeResult.status !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(encodeResult.status);
        const message = mteBase.getStatusDescription(encodeResult.status);
        throw new Error(`Error when MTE encoding data.\n${status}: ${message}`);
    }
    encoderState = encoder.saveStateB64();
    encoder.uninstantiate();
    encoder.destruct();
    return encodeResult.str;
}

/**
 * Use MTE to decrypt a uint8Array
 * - create mke decoder, restore it from saved mte state
 * - decode the array
 * - save new decoder state, destroy decoder
 * - return decrypted result
 */
export function mkeDecryptUint8Array(arr) {
    const decoder = MteMkeDec.fromdefault(mteWasm);
    decoder.restoreStateB64(decoderState);
    const decodeResult = decoder.decode(arr);
    if (decodeResult.status !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(decodeResult.status);
        const message = mteBase.getStatusDescription(decodeResult.status);
        throw new Error(`Error when MKE decrypting data.\n${status}: ${message}`);
    }
    decoderState = decoder.saveStateB64();
    decoder.uninstantiate();
    decoder.destruct();
    return decodeResult.arr;
}

/**
 * Uses the MkeDecoder to decode a base64 encoded payload and return it as string
 * @param str An MKE encoded value as a base64 encoded string.
 * @return A string.
 */
export function mkeDecryptString(str) {

    const decoder = MteMkeDec.fromdefault(mteWasm);
    decoder.restoreStateB64(decoderState);

    const decodeResult = decoder.decodeStrB64(str);
    // check if decode failed
    if (decodeResult.status !== MteStatus.mte_status_success) {
        const status = mteBase.getStatusName(decodeResult.status);
        const message = mteBase.getStatusDescription(decodeResult.status);
        throw new Error(`Error when MTE decoding data.\n${status}: ${message}`);
    }
    decoderState = decoder.saveStateB64();
    decoder.uninstantiate();
    decoder.destruct();
    return decodeResult.str;
}